rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,451 @@
+    let state = cast_pointer!(state, MQTTState);
+    return state.tx_id;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_state_progress_completion_status(_direction: u8) -> std::os::raw::c_int {
+    return 1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_alstate_progress(
+    _tx: *mut std::os::raw::c_void,
+    _direction: u8,
+) -> std::os::raw::c_int {
+    return 1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_get_logged(
+    _state: *mut std::os::raw::c_void,
+    tx: *mut std::os::raw::c_void,
+) -> u32 {
+    let tx = cast_pointer!(tx, MQTTTransaction);
+    return tx.logged.get();
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_tx_set_logged(
+    _state: *mut std::os::raw::c_void,
+    tx: *mut std::os::raw::c_void,
+    logged: u32,
+) {
+    let tx = cast_pointer!(tx, MQTTTransaction);
+    tx.logged.set(logged);
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_state_get_events(
+    tx: *mut std::os::raw::c_void,
+) -> *mut core::AppLayerDecoderEvents {
+    let tx = cast_pointer!(tx, MQTTTransaction);
+    return tx.events;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_state_get_event_info(
+    _event_name: *const std::os::raw::c_char,
+    _event_id: *mut std::os::raw::c_int,
+    _event_type: *mut core::AppLayerEventType,
+) -> std::os::raw::c_int {
+    return -1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_state_get_event_info_by_id(
+    _event_id: std::os::raw::c_int,
+    _event_name: *mut *const std::os::raw::c_char,
+    _event_type: *mut core::AppLayerEventType,
+) -> i8 {
+    return -1;
+}
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_state_get_tx_iterator(
+    _ipproto: u8,
+    _alproto: AppProto,
+    state: *mut std::os::raw::c_void,
+    min_tx_id: u64,
+    _max_tx_id: u64,
+    istate: &mut u64,
+) -> applayer::AppLayerGetTxIterTuple {
+    let state = cast_pointer!(state, MQTTState);
+    match state.tx_iterator(min_tx_id, istate) {
+        Some((tx, out_tx_id, has_next)) => {
+            let c_tx = unsafe { transmute(tx) };
+            let ires = applayer::AppLayerGetTxIterTuple::with_values(c_tx, out_tx_id, has_next);
+            return ires;
+        }
+        None => {
+            return applayer::AppLayerGetTxIterTuple::not_found();
+        }
+    }
+}
+
+// Parser name as a C style string.
+const PARSER_NAME: &'static [u8] = b"mqtt\0";
+
+export_tx_detect_flags_set!(rs_mqtt_set_tx_detect_flags, MQTTTransaction);
+export_tx_detect_flags_get!(rs_mqtt_get_tx_detect_flags, MQTTTransaction);
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_mqtt_register_parser() {
+    let default_port = CString::new("[1883]").unwrap();
+    let parser = RustParser {
+        name: PARSER_NAME.as_ptr() as *const std::os::raw::c_char,
+        default_port: default_port.as_ptr(),
+        ipproto: IPPROTO_TCP,
+        probe_ts: Some(rs_mqtt_probing_parser),
+        probe_tc: Some(rs_mqtt_probing_parser),
+        min_depth: 0,
+        max_depth: 16,
```

Comment:
Why magic 16 ?

Reply:
This was just the default from the template, as seen in other implementations as well. I'd be happy to change it, but I couldn't see any negative effects in my cases.

Reply:
@victorjulien any thoughts here ? Should we name it `SC_DEFAULT_PROTO_MAX_DEPTH` and keep value 16 ?